src/Security/AccessRightsPolicy.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace App\Security;
+
+use App\Entity\Activity;
+use App\Entity\User;
+use App\Repository\ActivityUserRepository;
+
+class AccessRightsPolicy
+{
+
+    /**
+     * @var ActivityUserRepository
+     */
+    private $repository;
+
+    public function __construct(ActivityUserRepository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    public function checkRightsToActivity(Activity $activity, User $user): bool
+    {
+        $userToActivity = $this->repository->findBy(array('activity' => $activity, 'user' => $user));
```

Comment:
Do the check `$activity->isPublic() === false) && $user !== $activity->getOwner()` before checking if user is invited/has applied. Most of the jobs will be public and the request to the database won't be needed.